### PR TITLE
*: Add blackbox probing for execution and duration of queries

### PR DESCRIPTION
**Pull request by brancz**
This change adds the `up` tool as a long running process running
continuously executing a set of defined queries monitoring their
ability to execute as well as their duration.

This marked as WIP as I want to get https://github.com/observatorium/up/pull/13 merged first.

@bwplotka @metalmatze @kakkoyun @squat @krasi-georgiev

### CODE REVIEW

**Comment by metalmatze** [comment]
Is this really a feature we want to have in upstream Observatorium? I seems to be pretty custom for our use case?
The blocking PR is merged so in general, we can move forward with this.

**Comment by brancz** [comment]
> Is this really a feature we want to have in upstream Observatorium? I seems to be pretty custom for our use case?

I don't feel too strongly about it, I'm happy to remove it from the base for now, I felt it might be useful for other setups, but maybe you're right and we can keep it to just the prod env for now.

File: components/up-job.libsonnet [hunk 1]
```diff
@@ -0,0 +1,79 @@
+local k = import 'ksonnet/ksonnet.beta.4/k.libsonnet';
+
+{
+  local up = self,
+
+  config:: {
+    name: error 'must provide name',
```
Comment (bwplotka) [inline components/up-job.libsonnet]: What's this? :scream:  Was not aware of this feature.

File: components/up-job.libsonnet [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  job:
+    local job = k.batch.v1.job;
+    local container = job.mixin.spec.template.spec.containersType;
+
+    local c =
+      container.new('observatorium-up', up.config.image) +
+      container.withArgs([
+        '--endpoint-write=' + up.config.writeEndpoint,
+        '--endpoint-read=' + up.config.readEndpoint,
+        '--period=1s',
```
Comment (bwplotka) [inline components/up-job.libsonnet]: :fire:

File: components/up-job.libsonnet [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      'app.kubernetes.io/component': 'test',
+    },
+  },
+
+  job:
+    local job = k.batch.v1.job;
+    local container = job.mixin.spec.template.spec.containersType;
+
+    local c =
+      container.new('observatorium-up', up.config.image) +
```
Comment (bwplotka) [inline components/up-job.libsonnet]: I am missing some flag for configuring what tests to run.. Is this hardcoded?

File: components/up-job.libsonnet [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      container.withArgs([
+        '--endpoint-write=' + up.config.writeEndpoint,
+        '--endpoint-read=' + up.config.readEndpoint,
+        '--period=1s',
+        '--duration=2m',
+        '--name=foo',
+        '--labels=bar="baz"',
+        '--latency=10s',
```
Comment (bwplotka) [inline components/up-job.libsonnet]: this flag means... timeout?

File: components/up-job.libsonnet [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  },
+
+  job:
+    local job = k.batch.v1.job;
+    local container = job.mixin.spec.template.spec.containersType;
+
+    local c =
+      container.new('observatorium-up', up.config.image) +
+      container.withArgs([
+        '--endpoint-write=' + up.config.writeEndpoint,
```
Comment (bwplotka) [inline components/up-job.libsonnet]: How to configure how often those tests are performed? :thinking:

File: environments/openshift/manifests/observatorium-template.yaml [hunk 6]
```diff
@@ -2062,6 +2062,133 @@ objects:
...
+      - "name": "Failed 4.2 upgrades"
+        "query": "count(((max by (_id) (sum_over_time((1+0*cluster_version{from_version=~\"4\\\\.1\\\\.\\\\d+\",version=~\"4\\\\.2\\\\.\\\\d+\",type=\"failure\"})[7d:15m]))) > 2) + on(_id) group_left(_blah) (topk by (_id) (1, 0*subscription_labels{email_domain!~\"redhat.com|(.*\\\\.|^)ibm.com\"})))"
+      - "name": "Clusters upgrading to 4.3"
+        "query": "count(((count by (_id) (count_over_time(cluster_version{from_version=~\"4\\\\.2\\\\.\\\\d+\",version=~\"4\\\\.3\\\\.\\\\d+\",type=\"updating\"}[7d])))*0+1) + on(_id) group_left(_blah) (topk by (_id) (1, 0*subscription_labels{email_domain!~\"redhat.com|(.*\\\\.|^)ibm.com\"})))"
+      - "name": "Failed 4.3 upgrades"
+        "query": "count(((max by (_id) (sum_over_time((1+0*cluster_version{from_version=~\"4\\\\.2\\\\.\\\\d+\",version=~\"4\\\\.3\\\\.\\\\d+\",type=\"failure\"})[7d:15m]))) > 2) + on(_id) group_left(_blah) (topk by (_id) (1, 0*subscription_labels{email_domain!~\"redhat.com|(.*\\\\.|^)ibm.com\"})))"
```
Comment (bwplotka) [inline environments/openshift/manifests/observatorium-template.yaml]: Good work on those!

Again, the question is how we ensure this is synchronized... with @smarterclayton  (:  :hugs:

File: environments/openshift/obs.jsonnet [hunk 7]
```diff
@@ -486,6 +487,30 @@ local cqf = (import '../../components/cortex-query-frontend.libsonnet');
...
+      version: 'master-2020-03-25-6d4f944',
+      image: 'quay.io/observatorium/up:' + cfg.version,
+      queryConfig: (import 'queries.libsonnet'),
+      resources: {
+        requests: {
+          cpu: '5m',
+          memory: '10Mi',
+        },
+        limits: {
+          cpu: '20m',
```
Comment (bwplotka) [inline environments/openshift/obs.jsonnet]: This can really affect test runs from client side. I would give 1CPU, we are not short on CPU I think

File: environments/openshift/obs.jsonnet [hunk 8]
```diff
@@ -486,6 +487,30 @@ local cqf = (import '../../components/cortex-query-frontend.libsonnet');
...
+          memory: '50Mi',
```
Comment (bwplotka) [inline environments/openshift/obs.jsonnet]: same here... `200Mi`?

File: components/up-job.libsonnet (hunk 4 above)
Comment (squat) [reply components/up-job.libsonnet]: this is the maximum allowed diff between writing and reading a sample

File: environments/openshift/obs.jsonnet (hunk 7 above)
Comment (brancz) [reply environments/openshift/obs.jsonnet]: Right now all this thing does is to an HTTP request.

File: components/up-job.libsonnet (hunk 1 above)
Comment (brancz) [reply components/up-job.libsonnet]: one of my fav features! that and assert for testing are the most underused and underappreciated features

File: components/up-job.libsonnet (hunk 3 above)
Comment (brancz) [reply components/up-job.libsonnet]: sorry the git diff makes this hard to follow, this file was just renamed from `up.libsonnet` to `up-job.libsonnet`, separating the job that we use in our e2e tests, and the long running process we run in prod

File: components/up-job.libsonnet (hunk 5 above)
Comment (brancz) [reply components/up-job.libsonnet]: it's defined by `duration=2m` meaning we will probe for a total of 2 minutes in 1s periods and 90% write must have a read-after-write latency of less than 10s

Comment (bwplotka) [reply components/up-job.libsonnet]: Ok, but how often the test is running? Constantly?

Comment (brancz) [reply components/up-job.libsonnet]: every 1 second for 2 minutes, so ~120 times
